---
title: "HDAnalyzeR"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{HDAnalyzeR}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## What is HDAnalyzeR?

HDAnalyzeR is an R package developed to streamline and enhance proteomics analysis, particularly for biomarker discovery from blood plasma samples. It is developed by the Human Disease Blood Atlas group. This vignette will guide you through the essential steps to use the package from data loading to biomarker identification. Let's get started by loading the package!

```{r setup}
library(HDAnalyzeR)
```

## How to get Help?

If you need help with the package, you can use the `?` operator to access the documentation for a specific function. For example, to access the documentation for the `hd_initialize()` function, you can run:

```{r, eval = FALSE}
?hd_initialize
```

All package functions can be found in the [HDAnalyzeR Function Reference](https://hda1472.github.io/HDAnalyzeR/reference/index.html) page. For whole example workflows you can visit any of the package articles.

This document introduces you to HDAnalyzeR's basic set of tools, and shows you how to analyze and identify biomarkers in a artificial dataset of cancer blood plasma samples.

## Loading the Data

First, we load the package's `example_data` and `example_metadata`.

```{r}
head(example_data)
```

```{r}
head(example_metadata)
```

We will initialize the HDAnalyzeR object with the data and metadata. The `hd_initialize()` function will create an object that will be used throughout the analysis. We will specify the sample ID, variable name, and value name in order to correctly load and widen the data. In this case, the data is in long format, so we set `is_wide = FALSE`.

```{r, results = 'hide', message = FALSE, warning = FALSE}
hd_obj <- hd_initialize(dat = example_data, 
                        metadata = example_metadata, 
                        is_wide = FALSE, 
                        sample_id = "DAid",
                        var_name = "Assay",
                        value_name = "NPX")
```

```{r}
# See the data transformed into a wide format!
head(hd_obj$data)
```

> üìì In real-world scenarios, you would load your own data and metadata files instead of using the example dataset.

> ‚ö†Ô∏è It is strongly recommended to use `hd_initialize()` function to initialize an HDAnalyzeR object before starting any analysis. This initialization step ensures that your input data and metadata are correctly formatted and compatible with the package. While it is possible to use HDAnalyzeR functions on data outside of an HDAnalyzeR object, users must ensure that the data adheres to the required structure: the first column must contain sample IDs, followed by numeric columns in wide format. Additionally, the metadata must include the same set of sample IDs. Proper initialization or careful adherence to these requirements is crucial for accurate and efficient analysis.

## Quality Control (QC)

`hd_qc_summary()` provides a comprehensive summary of the input data and metadata. 

Regarding the data, it will check the column types, calculate the percentage of NAs in each column and row and plot their distributions, calculate variable-variable (in this case protein-protein) correlations, and create a heatmap of these correlations. Users can also specify the threshold for reporting variable-variable correlations.

Regarding the metadata, it will check the column types and calculate the percentage of NAs in each column and row exactly as before.  It will also create summary visualizations for the metadata variables like "Sex" and "Age".

Let's run the QC summary and take a look at some results!

```{r}
qc_data <- hd_qc_summary(hd_obj, 
                         variable = "Disease", 
                         palette = list(Disease = "cancers12", Sex = "sex"), 
                         cor_threshold = 0.7,
                         verbose = FALSE)

qc_data$data_summary$na_col_hist
qc_data$data_summary$cor_heatmap
qc_data$metadata_summary$Sex
qc_data$metadata_summary$Age
```

## Imputation and Dimensionality Reduction

As we saw from the QC results, the data contains NAs and other issues that need to be addressed. We will impute missing values using K-nearest neighbors (KNN) with 5 neighbors via `hd_impute_knn()`.

```{r}
hd_obj <- hd_impute_knn(hd_obj, k = 5, verbose = FALSE)
head(hd_obj$data)
```

After imputation, we will run Principal Component Analysis (PCA) via `hd_auto_pca()` and Uniform Manifold Approximation and Projection (UMAP) via `hd_auto_umap()` to check for outliers, batch effects, and other potential issues. These are wrapper functions that will run the necessary steps to perform PCA and UMAP, including scaling the data and coloring the plots by metadata variables.

```{r}
pca_res <- hd_auto_pca(hd_obj, 
                       components = 10, 
                       plot_color = "Disease", 
                       plot_palette = "cancers12")

head(pca_res$pca_res)
pca_res$pca_plot
pca_res$pca_loadings_plot
pca_res$pca_variance_plot
```

Let's color UMAP by Sex instead of Disease to see if there are any sex related differences.

```{r}
umap_res <- hd_auto_umap(hd_obj, 
                         plot_color = "Sex", 
                         plot_palette = "sex")

head(umap_res$umap_res)
umap_res$umap_plot
```

## Biomarker Identification

### Differential Expression Analysis

We will run a differential expression analysis to identify potential biomarkers. We will use `hd_de_limma()` so that we will be able to correct also for Sex and Age. This method will help us pinpoint proteins that are significantly different between conditions. We will run the analysis 3 times for different case classes against all other classes each time. We will also run `hd_plot_volcano()` to visualize the results. We will present only the results for Acute Myeloid Leukemia (AML).

```{r}
de_res_aml <- hd_de_limma(hd_obj,
                          variable = "Disease",
                          case = "AML",
                          correct = c("Sex", "Age")) |> 
  hd_plot_volcano(logfc_lim = 1)

head(de_res_aml$de_res)
de_res_aml$volcano_plot
```

```{r, results = 'hide', message = FALSE, warning = FALSE}
de_res_crc <- hd_de_limma(hd_obj,
                          variable = "Disease",
                          case = "CRC",
                          correct = c("Sex", "Age"))
```

```{r, results = 'hide', message = FALSE, warning = FALSE}
de_res_lungc <- hd_de_limma(hd_obj,
                            variable = "Disease",
                            case = "LUNGC",
                            correct = c("Sex", "Age"))
```

We can also summarize the results via `hd_plot_de_summary()`. In order to use this function we need to store all the results in a list.

```{r, results = 'hide', message = FALSE, warning = FALSE}
de_res <- list("AML" = de_res_aml, 
               "CRC" = de_res_crc, 
               "LUNGC" = de_res_lungc)

de_summary <- hd_plot_de_summary(de_res, class_palette = "cancers12")
```

```{r}
de_summary$de_barplot
de_summary$upset_plot_up
de_summary$upset_plot_down
```

### Machine Learning Classification Model

In addition to differential expression analysis, we will use a regularized machine learning classification model (LASSO) to identify significant features. This model will help us understand which protein features are most predictive of the conditions being studied. 

We will start by creating the data split with `hd_split_data()` so we don't have any data leakage between train and test data and then we will run the model with `hd_model_rreg()`. Once again, we will run the analysis 3 times but we will present only the results for AML.

```{r, warning = FALSE, message = FALSE}
hd_split <- hd_split_data(hd_obj, variable = "Disease")

lasso_res_aml <- hd_model_rreg(hd_split, 
                               variable = "Disease",
                               case = "AML",
                               mixture = 1,
                               grid_size = 5,
                               palette = "cancers12",
                               verbose = FALSE)

lasso_res_aml$metrics
lasso_res_aml$roc_curve
lasso_res_aml$probability_plot
lasso_res_aml$feat_imp_plot
```

```{r, results = 'hide', message = FALSE, warning = FALSE}
lasso_res_crc <- hd_model_rreg(hd_split, 
                               variable = "Disease",
                               case = "CRC",
                               mixture = 1,
                               grid_size = 5,
                               palette = "cancers12",
                               verbose = FALSE)
```

```{r, results = 'hide', message = FALSE, warning = FALSE}
lasso_res_lungc <- hd_model_rreg(hd_split, 
                                 variable = "Disease",
                                 case = "LUNGC",
                                 mixture = 1,
                                 grid_size = 5,
                                 palette = "cancers12",
                                 verbose = FALSE)
```

We can get a visual summary of the results via `hd_plot_model_summary()`. In order to use this function we need to store the results in a list as before.

```{r, results = 'hide'}
lasso_res <- list("AML" = lasso_res_aml, 
                  "CRC" = lasso_res_crc, 
                  "LUNGC" = lasso_res_lungc)

features_summary <- hd_plot_model_summary(lasso_res, class_palette = "cancers12")
```

```{r}
features_summary$metrics_barplot
features_summary$features_barplot
features_summary$upset_plot_features
```

## One step further

The final step involves performing a pathway enrichment analysis to get more biological information about the features. In this example, we will run an Over-Representation Analysis for the up-regulated proteins from differential expression, that are also identified as features by the ML model. We will use the Gene Ontology (GO) database and show the results only for AML. We will use `hd_ora()` and `hd_plot_ora()` to perform the analysis and visualize its results.

```{r, message = FALSE, warning = FALSE}
# Extract the proteins identified by both DE and Lasso
de_proteins <- de_res_aml$de_res |> 
  dplyr::filter(adj.P.Val < 0.05) |> 
  dplyr::pull(Feature)

lasso_proteins <- lasso_res_aml$features |> 
  dplyr::filter(Scaled_Importance > 0) |> 
  dplyr::pull(Feature)

intersect_proteins <- intersect(de_proteins, lasso_proteins)

# Perform ORA with GO database and visualize results
enrichment <- hd_ora(intersect_proteins, 
                     database = "GO", 
                     ontology = "BP", 
                     pval_lim = 0.2)  # We will use a higher p-value threshold for this example
enrichment_plots <- hd_plot_ora(enrichment)

enrichment_plots$dotplot
enrichment_plots$treeplot
enrichment_plots$cnetplot
```

> üìì Remember that these data are a dummy-dataset with artificial data and the results in this guide should not be interpreted as real results. The purpose of this vignette is to show you how to use the package and its functions.

```{r session-info}
sessionInfo()
```