---
title: "solid_cancers_transcriptomics_case_study"
output: html_document
---

# Case Study 2: Transcriptomic Signatures Across Solid Tumors in CPTAC Datasets

## Set-up & Data

Download and preprocess data using `get_CPTAC_cancer_TPMs.R` script them and store the resulting data in the `data/` folder.

Reference: Edwards NJ, Oberti M, Thangudu RR, Cai S, McGarvey PB, Jacob S, Madhavan S, Ketchum KA. The CPTAC Data Portal: A Resource for Cancer Proteomics Research. J Proteome Res. 2015 Jun 5;14(6):2707-13. doi: 10.1021/pr501254j. Epub 2015 May 4. PMID: 25873244.

```{r}
library(dplyr)
library(readr)
library(ggplot2)
library(SummarizedExperiment)
library(HDAnalyzeR)
library(patchwork)

# Read data
data_kidney <- readRDS("data/all_CPTAC_KIAD.rds")
data_lung <- readRDS("data/all_CPTAC_LUAD.rds")
data_endometrial <- readRDS("data/all_CPTAC_UCEC.rds")

# Read metadata
meta_kidney <- hd_import_data("data/sample_sheet_KIAD.tsv")
meta_lung <- read_tsv("data/sample_sheet_CPTAC_LUAD.txt")
meta_endometrial <- read_tsv("data/sample_sheet_CPTAC_UCEC.txt")
```

## Palettes

```{r}
palette_sample_type <- c("Tumor" = "#F3717A",
                         "Normal" = "#D6DA90")


palette_cancer <- c("Kidney" = "#F9A265",
                    "Lung" = "#6BA592", 
                    "Endometrial" = "#F8BCD7")
```

## Preprocessing

Extract the TPM data from the object for all 3 cancers.

```{r}
df_kidney <- 
  assay(data_kidney, "tpm_unstrand") |> 
  as.data.frame()

df_lung <- 
  assay(data_lung, "tpm_unstrand") |> 
  as.data.frame()

df_endometrial <- 
  assay(data_endometrial, "tpm_unstrand") |> 
  as.data.frame()
```

We then only keep the TPM values for coding genes based on the HPA database (https://www.proteinatlas.org/humanproteome/tissue/data#normal_tissues_ihc). We will log transform the data as well by (log2(TPM + 1)).

```{r}
hpa_rna <- read_tsv("normal_tissue.tsv")

coding_genes <- 
  hpa_rna |> 
  distinct(Gene) |> 
  pull()

gene_mapping <- 
  tibble(gene = rownames(df_kidney)) |> 
  mutate(gene_no_suffix = sub("\\..*", "", gene)) |> 
  left_join(hpa_rna |> 
              distinct(Gene, `Gene name`), by = c("gene_no_suffix" = "Gene"))

parse_data <- function(df) {
  
  df |>
    as_tibble(rownames = "gene") |>
    
    # Exclude genes with _PAR_Y suffix
    filter(!str_ends(gene, "_PAR_Y")) |>
    
    # Split Ensembl gene ID and version
    mutate(gene_split = gene) |> 
    separate(gene_split, into = c("gene_no_suffix", "version"), sep = "\\.", convert = TRUE) |>
    
    # Keep only genes that are in your list of coding genes
    filter(gene_no_suffix %in% coding_genes) |>
    
    # Map Ensembl ID to gene symbol (HGNC name)
    left_join(gene_mapping, by = c("gene", "gene_no_suffix")) |>
    
    # Keep the lowest ensembl id for genes names with multiple mappings
    group_by(`Gene name`) |>
    slice_min(order_by = gene_no_suffix, with_ties = FALSE) |>
    ungroup() |>
    
    # Clean up and reshape
    select(-gene_no_suffix, -version, -gene) |>
    relocate(gene = `Gene name`) |>
    pivot_longer(
      cols = -gene,
      names_to = "sample_id",
      values_to = "tpm"
    ) |> 
    mutate(tpm = log(tpm + 1 , base = exp(1)))

}

kidney <- parse_data(df_kidney)
lung <- parse_data(df_lung)
endometrial <- parse_data(df_endometrial)
```


We will also extract some metadata variables: age and sex.

```{r}
kidney_age_sex <- tibble(
  sample_id = data_kidney$sample,
  sex = data_kidney$gender,
  age = data_kidney$age_at_diagnosis
) |>
  mutate(age = as.numeric(age) / 365)

lung_age_sex <- tibble(
  sample_id = data_lung$sample,
  sex = data_lung$gender,
  age = data_lung$age_at_diagnosis
) |>
  mutate(age = as.numeric(age) / 365)

endometrial_age_sex <- tibble(
  sample_id = data_endometrial$sample,
  sex = data_endometrial$gender,
  age = data_endometrial$age_at_diagnosis
) |>
  mutate(age = as.numeric(age) / 365)
```

Finally, we will process the metadata by renaming some variables, filtering the samples based on the IDs in the data and joining the age and sex information.

```{r}
# Function to parse metadata
parse_meta <- function(meta,
                       data,
                       age_sex,
                       cancer) {
  meta |>
    filter(sample.submitter_id %in% data$sample_id) |>
    distinct(sample.submitter_id, sample_type) |>
    mutate(
      cancer = cancer,
      sample_type = sub(";.*", "", sample_type),
      sample_type_short = recode(
        sample_type,
        "Primary Tumor" = "Tumor",
        "Solid Tissue Normal" = "Normal"
      ),
      sample_combined = paste(cancer, sample_type_short)
    ) |>
    dplyr::rename(sample_id = sample.submitter_id) |> 
    left_join(age_sex, by = "sample_id")
}

# Parse metadata for all cancers
meta_simple_kidney <- parse_meta(
  meta = meta_kidney,
  data = kidney,
  age_sex = kidney_age_sex,
  cancer = "Kidney"
)

meta_simple_lung <- parse_meta(
  meta = meta_lung,
  data = lung,
  age_sex = lung_age_sex,
  cancer = "Lung"
)

meta_simple_endometrial <- parse_meta(
  meta = meta_endometrial,
  data = endometrial,
  age_sex = endometrial_age_sex,
  cancer = "Endometrial"
)
```

## Initialize HDAobject

```{r}
hd_kidney <-
  hd_initialize(
    kidney,
    meta_simple_kidney,
    sample_id = "sample_id",
    var_name = "gene",
    value_name = "tpm"
  )

hd_lung <-
  hd_initialize(
    lung,
    meta_simple_lung,
    sample_id = "sample_id",
    var_name = "gene",
    value_name = "tpm"
  )

hd_endometrial <-
  hd_initialize(
    endometrial,
    meta_simple_endometrial,
    sample_id = "sample_id",
    var_name = "gene",
    value_name = "tpm"
  )
```

## PCA

We will run a PCA analysis on the data to check if the data contain any outliers or cluster in an unexpected way. The analysis will be run for each cancer separately.

```{r}
# Kidney
pca_kidney <- hd_pca(hd_kidney)
pca_plot_kidney <- 
  pca_kidney |>
  hd_plot_dim(hd_kidney, "PC1", "PC2", color = "sample_type_short", palette = palette_sample_type, axis_variance = TRUE) 

# Lung
pca_lung <- hd_pca(hd_lung)
pca_plot_lung <- 
  pca_lung |>
  hd_plot_dim(hd_lung, "PC1", "PC2", color = "sample_type_short", palette = palette_sample_type, axis_variance = TRUE) 

# Endometrial
pca_endometrial <- hd_pca(hd_endometrial)
pca_plot_endometrial <- 
  pca_endometrial |>
  hd_plot_dim(hd_endometrial, "PC1", "PC2", color = "sample_type_short", palette = palette_sample_type,  axis_variance = TRUE) 

# Combined plots
combined_pca_plots <- 
  pca_plot_kidney$pca_plot | 
  pca_plot_lung$pca_plot | 
  pca_plot_endometrial$pca_plot 
combined_pca_plots <- combined_pca_plots +  plot_layout(guides = "collect")
```

We will also check the variance of the principal components for each cancer to see which transcripts are driving the variance in each case.

```{r}
# Kidney
pca_variance_kidney <- 
  pca_kidney |>   
  hd_plot_pca_loadings(displayed_pcs = 1, displayed_features = 10)

# Lung
pca_variance_lung <- 
  pca_lung |>   
  hd_plot_pca_loadings(displayed_pcs = 1, displayed_features = 10)

# Endometrial
pca_variance_endometrial <- 
  pca_endometrial |>   
  hd_plot_pca_loadings(displayed_pcs = 1, displayed_features = 10)

combined_variance_plots <- 
  pca_variance_kidney$pca_loadings_plot |
  pca_variance_lung$pca_loadings_plot | 
  pca_variance_endometrial$pca_loadings_plot 
combined_variance_plots <- combined_variance_plots +  plot_layout(guides = "collect")
```

Finally, we will plot the top two features that drive the variance in each case as boxplots.

```{r}
# Kidney
top_features_kidney <- 
  pca_variance_kidney$pca_loadings |> 
  filter(component == "PC1") |> 
  arrange(-abs(value)) |> 
  head(2) |> 
  pull(terms)

top_features_kidney_plot <- 
  hd_plot_feature_boxplot(hd_kidney, 
                        features = top_features_kidney, 
                        variable = "sample_type_short",
                        palette = palette_sample_type, 
                        points = T)

# Lung
top_features_lung <- 
  pca_variance_lung$pca_loadings |> 
  filter(component == "PC1") |> 
  arrange(-abs(value)) |> 
  head(2) |> 
  pull(terms)

top_features_lung_plot <- 
  hd_plot_feature_boxplot(hd_lung, 
                        features = top_features_lung, 
                        variable = "sample_type_short",
                        palette = palette_sample_type, 
                        points = T)

# Endometrial
top_features_endometrial <- 
  pca_variance_endometrial$pca_loadings |> 
  filter(component == "PC1") |> 
  arrange(-abs(value)) |> 
  head(2) |> 
  pull(terms)

top_features_endometrial_plot <- 
  hd_plot_feature_boxplot(hd_endometrial, 
                        features = top_features_endometrial, 
                        variable = "sample_type_short",
                        palette = palette_sample_type, 
                        points = T)


combined_pca_boxplots_plots <- 
  top_features_kidney_plot | 
  top_features_lung_plot |
  top_features_endometrial_plot 
combined_pca_boxplots_plots <- combined_pca_boxplots_plots +  plot_layout(guides = "collect")
```


## DE

Then, we will run a DE analysis on the data again for each cancer separately and plot the results as volcano plots.

```{r}
de_kidney <- hd_de_limma(hd_kidney, 
                        variable = "sample_type_short", 
                        case = "Tumor", 
                        control = "Normal",
                        correct = c("age", "sex")
                        )

de_lung <- hd_de_limma(hd_lung, 
                        variable = "sample_type_short", 
                        case = "Tumor", 
                        control = "Normal", 
                        correct = c("age", "sex")
                        )

de_endometrial <- hd_de_limma(hd_endometrial, 
                        variable = "sample_type_short", 
                        case = "Tumor", 
                        control = "Normal", 
                        correct = "age"
                        )
```

```{r}
volcano_kidney <- 
  de_kidney |> 
  hd_plot_volcano(report_nproteins = FALSE,
                  top_up_prot = 3,
                  top_down_prot = 3)

volcano_lung <- 
  de_lung |> 
  hd_plot_volcano(report_nproteins = FALSE,
                  top_up_prot = 3,
                  top_down_prot = 3)


volcano_endometrial <- 
  de_endometrial |> 
  hd_plot_volcano(report_nproteins = FALSE,
                  top_up_prot = 3,
                  top_down_prot = 3)


combined_volcano_plots <- 
  volcano_kidney$volcano_plot | 
  volcano_lung$volcano_plot |
  volcano_endometrial$volcano_plot 
combined_volcano_plots <- combined_volcano_plots +  plot_layout(guides = "collect")
```

We will plot the most up- and the most down-regulated transcripts as boxplots to see their difference between the cancer and control samples.
  
```{r}
# Kidney
top_features_de_kidney <- 
  de_kidney$de_res |> 
  slice_max(logFC, n = 1) |>         
  bind_rows(slice_min(de_kidney$de_res, logFC, n = 1)) |> 
  pull(Feature)

top_features_de_kidney_plot <- 
  hd_plot_feature_boxplot(hd_kidney, 
                          features = top_features_de_kidney, 
                          variable = "sample_type_short",
                          palette = palette_sample_type, 
                          points = T)

# Lung
top_features_de_lung <- 
  de_lung$de_res |> 
  slice_max(logFC, n = 1) |>         
  bind_rows(slice_min(de_lung$de_res, logFC, n = 1)) |> 
  pull(Feature)

top_features_de_lung_plot <- 
  hd_plot_feature_boxplot(hd_lung, 
                          features = top_features_de_lung, 
                          variable = "sample_type_short",
                          palette = palette_sample_type, 
                          points = T)

# Endometrial
top_features_de_endometrial <- 
  de_endometrial$de_res |> 
  slice_max(logFC, n = 1) |>         
  bind_rows(slice_min(de_endometrial$de_res, logFC, n = 1)) |> 
  pull(Feature)

top_features_de_endometrial_plot <- 
  hd_plot_feature_boxplot(hd_endometrial, 
                          features = top_features_de_endometrial, 
                          variable = "sample_type_short",
                          palette = palette_sample_type, 
                          points = T)

combined_de_boxplots <- 
  top_features_de_kidney_plot | 
  top_features_de_lung_plot |
  top_features_de_endometrial_plot
combined_de_boxplots <- combined_de_boxplots +  plot_layout(guides = "collect")
```

Last but not least, we will plot the DE results for all 3 analyses as summary upset plots, one for up- and one for down-regulated transcripts.

```{r}
de_summary_res <- hd_plot_de_summary(list("Kidney" = de_kidney, 
                                          "Lung" = de_lung, 
                                          "Endometrial" = de_endometrial),
                                     variable = "sample_type_short",
                                     class_palette = palette_cancer)

de_summary_res$upset_plot_up 

de_summary_res$upset_plot_down
```