<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Over-representation analysis — hd_ora • HDAnalyzeR</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Over-representation analysis — hd_ora"><meta name="description" content="hd_ora() performs over-representation analysis (ORA) using the clusterProfiler package."><meta property="og:description" content="hd_ora() performs over-representation analysis (ORA) using the clusterProfiler package."><meta property="og:image" content="https://kantonopoulos.github.io/HDAnalyzeR/logo.png"></head><body><p>









  .github/pkgdown/extra.css

  

  </p>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">HDAnalyzeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/HDAnalyzeR.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/before_using_hdanalyzer.html">Data Preparation for HDAnalyzeR: What You Need Before Using the Package</a></li>
    <li><a class="dropdown-item" href="../articles/data_preprocessing.html">Exploratory Data Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/data_imputation.html">Data Imputation</a></li>
    <li><a class="dropdown-item" href="../articles/differential_expression.html">Differential Expression Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/classification.html">Machine Learning Models</a></li>
    <li><a class="dropdown-item" href="../articles/post_analysis.html">Post Analysis: Pathway Enrichment</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kantonopoulos/HDAnalyzeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Over-representation analysis</h1>
      <small class="dont-index">Source: <a href="https://github.com/kantonopoulos/HDAnalyzeR/blob/main/R/enrichment_analysis.R" class="external-link"><code>R/enrichment_analysis.R</code></a></small>
      <div class="d-none name"><code>hd_ora.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>hd_ora()</code> performs over-representation analysis (ORA) using the clusterProfiler package.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">hd_ora</span><span class="op">(</span></span>
<span>  <span class="va">gene_list</span>,</span>
<span>  database <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO"</span>, <span class="st">"Reactome"</span>, <span class="st">"KEGG"</span><span class="op">)</span>,</span>
<span>  ontology <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BP"</span>, <span class="st">"CC"</span>, <span class="st">"MF"</span>, <span class="st">"ALL"</span><span class="op">)</span>,</span>
<span>  background <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pval_lim <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-gene-list">gene_list<a class="anchor" aria-label="anchor" href="#arg-gene-list"></a></dt>
<dd><p>A character vector containing the gene names. These can be differentially expressed proteins or selected protein features from classification models.</p></dd>


<dt id="arg-database">database<a class="anchor" aria-label="anchor" href="#arg-database"></a></dt>
<dd><p>The database to perform the ORA. It can be either "GO", "KEGG", or "Reactome".</p></dd>


<dt id="arg-ontology">ontology<a class="anchor" aria-label="anchor" href="#arg-ontology"></a></dt>
<dd><p>The ontology to use when database = "GO". It can be "BP" (Biological Process), "CC" (Cellular Component), "MF" (Molecular Function), or "ALL". In the case of KEGG and Reactome, this parameter is ignored.</p></dd>


<dt id="arg-background">background<a class="anchor" aria-label="anchor" href="#arg-background"></a></dt>
<dd><p>A character vector containing the background genes or a string with the name of the background gene list to use (use <code><a href="hd_show_backgrounds.html">hd_show_backgrounds()</a></code> to see available lists). If NULL, the full proteome is used as background.</p></dd>


<dt id="arg-pval-lim">pval_lim<a class="anchor" aria-label="anchor" href="#arg-pval-lim"></a></dt>
<dd><p>The p-value threshold to consider a term as significant in the enrichment analysis.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the results of the ORA.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>To perform the ORA, <code>clusterProfiler</code> package is used.
The <code>qvalueCutoff</code> is set to 1 by default to prioritize filtering by adjusted
p-values (p.adjust). This simplifies the workflow by ensuring a single, clear
significance threshold based on the false discovery rate (FDR). While q-values
are not used for filtering by default, they are still calculated and included
in the results for users who wish to apply additional criteria.
For more information, please refer to the <code>clusterProfiler</code> documentation.</p>
<p>If you want to learn more about ORA, please refer to the following publications:</p><ul><li><p>Chicco D, Agapito G. Nine quick tips for pathway enrichment analysis. PLoS Comput Biol. 2022 Aug 11;18(8):e1010348. doi: 10.1371/journal.pcbi.1010348. PMID: 35951505; PMCID: PMC9371296. https://pmc.ncbi.nlm.nih.gov/articles/PMC9371296/</p></li>
<li><p>https://yulab-smu.top/biomedical-knowledge-mining-book/enrichment-overview.html#gsea-algorithm</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Initialize an HDAnalyzeR object</span></span></span>
<span class="r-in"><span><span class="va">hd_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="hd_initialize.html">hd_initialize</a></span><span class="op">(</span><span class="va">example_data</span>, <span class="va">example_metadata</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run differential expression analysis for AML vs all others</span></span></span>
<span class="r-in"><span><span class="va">de_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="hd_de_limma.html">hd_de_limma</a></span><span class="op">(</span><span class="va">hd_object</span>, case <span class="op">=</span> <span class="st">"AML"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Extract the up-regulated proteins for AML</span></span></span>
<span class="r-in"><span><span class="va">sig_up_proteins_aml</span> <span class="op">&lt;-</span> <span class="va">de_results</span><span class="op">$</span><span class="va">de_res</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">|&gt;</span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">Feature</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Perform ORA with `GO` database and `BP` ontology</span></span></span>
<span class="r-in"><span><span class="va">enrichment</span> <span class="op">&lt;-</span> <span class="fu">hd_ora</span><span class="op">(</span><span class="va">sig_up_proteins_aml</span>, database <span class="op">=</span> <span class="st">"GO"</span>, ontology <span class="op">=</span> <span class="st">"BP"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No background gene list provided. For meaningful enrichment results, it is recommended to specify a relevant background list of genes (e.g., the full proteome or a set of genes that could be impacted in your experiment). The absence of a background may lead to misleading results in the over-representation analysis (ORA).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:1 mapping between keys and columns</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>coercing argument of type 'double' to logical</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Access the results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">enrichment</span><span class="op">$</span><span class="va">enrichment</span><span class="op">@</span><span class="va">result</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    ID                          Description GeneRatio   BgRatio</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0050900 GO:0050900                  leukocyte migration      7/22 395/18805</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0050926 GO:0050926    regulation of positive chemotaxis      3/22  27/18805</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0071674 GO:0071674           mononuclear cell migration      5/22 230/18805</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0045785 GO:0045785 positive regulation of cell adhesion      6/22 499/18805</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0001666 GO:0001666                  response to hypoxia      5/22 316/18805</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0036293 GO:0036293  response to decreased oxygen levels      5/22 330/18805</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            RichFactor FoldEnrichment    zScore       pvalue     p.adjust</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0050900 0.01772152       15.14787  9.725607 2.222549e-07 0.0002611495</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0050926 0.11111111       94.97475 16.723315 3.991584e-06 0.0022807251</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0071674 0.02173913       18.58202  9.181680 5.823128e-06 0.0022807251</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0045785 0.01202405       10.27783  7.188758 1.760247e-05 0.0051707246</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0001666 0.01582278       13.52488  7.684481 2.702828e-05 0.0063516447</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0036293 0.01515152       12.95110  7.495950 3.326012e-05 0.0065134398</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                  qvalue                         geneID Count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0050900 0.0001586198 100/566/9048/25/2683/199/30817     7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0050926 0.0013852915                   566/285/9048     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0071674 0.0013852915           566/9048/25/2683/199     5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0045785 0.0031406506      100/566/54518/9289/25/199     6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0001666 0.0038579306         100/285/51129/405/1386     5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0036293 0.0039562036         100/285/51129/405/1386     5</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># With a background gene list</span></span></span>
<span class="r-in"><span><span class="va">enrichment</span> <span class="op">&lt;-</span> <span class="fu">hd_ora</span><span class="op">(</span><span class="va">sig_up_proteins_aml</span>,</span></span>
<span class="r-in"><span>                     database <span class="op">=</span> <span class="st">"GO"</span>,</span></span>
<span class="r-in"><span>                     ontology <span class="op">=</span> <span class="st">"BP"</span>,</span></span>
<span class="r-in"><span>                     background <span class="op">=</span> <span class="st">"olink_explore_ht"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:1 mapping between keys and columns</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'select()' returned 1:1 mapping between keys and columns</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>1.31% of input gene IDs are fail to map...</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>coercing argument of type 'double' to logical</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Access the results</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">enrichment</span><span class="op">$</span><span class="va">enrichment</span><span class="op">@</span><span class="va">result</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                    ID                                   Description GeneRatio</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0050900 GO:0050900                           leukocyte migration      7/22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0007186 GO:0007186  G protein-coupled receptor signaling pathway      7/22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0050926 GO:0050926             regulation of positive chemotaxis      3/22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0001666 GO:0001666                           response to hypoxia      5/22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0040012 GO:0040012                      regulation of locomotion      9/22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0060135 GO:0060135 maternal process involved in female pregnancy      3/22</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>             BgRatio RichFactor FoldEnrichment   zScore       pvalue   p.adjust</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0050900 209/4805 0.03349282       7.315137 6.330362 2.600575e-05 0.02554723</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0007186 244/4805 0.02868852       6.265835 5.725254 7.066822e-05 0.02554723</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0050926  19/4805 0.15789474      34.485646 9.917719 7.700889e-05 0.02554723</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0001666 115/4805 0.04347826       9.496047 6.253772 1.368573e-04 0.02554723</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0040012 479/4805 0.01878914       4.103720 4.854781 1.368986e-04 0.02554723</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0060135  23/4805 0.13043478      28.488142 8.961221 1.390836e-04 0.02554723</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                qvalue                                geneID Count</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0050900 0.02125936        100/566/9048/25/2683/199/30817     7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0007186 0.02125936         100/566/9289/181/25/976/30817     7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0050926 0.02125936                          566/285/9048     3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0001666 0.02125936                100/285/51129/405/1386     5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0040012 0.02125936 100/566/9289/285/9048/25/59/51742/199     9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GO:0060135 0.02125936                         285/181/59272     3</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Konstantinos Antonopoulos, Maria Bueno Alvez, Mathias Uhlen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Provided without <strong>any warranty</strong>.</p>
</div>

    </footer></div>





  
</body></html>

