<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>HDAnalyzeR • HDAnalyzeR</title>
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="HDAnalyzeR">
</head>
<body>
<p>








  pkgdown/extra.css

  

  </p>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">HDAnalyzeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/HDAnalyzeR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/before_using_hdanalyzer.html">Data Preparation for HDAnalyzeR: What You Need Before Using the Package</a></li>
    <li><a class="dropdown-item" href="../articles/data_preprocessing.html">Exploratory Data Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/data_imputation.html">Data Imputation</a></li>
    <li><a class="dropdown-item" href="../articles/differential_expression.html">Differential Expression Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/classification.html">Machine Learning Models</a></li>
    <li><a class="dropdown-item" href="../articles/post_analysis.html">Post Analysis: Pathway Enrichment &amp; Automated Literature Search</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/kantonopoulos/HDAnalyzeR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>HDAnalyzeR</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/kantonopoulos/HDAnalyzeR/blob/website/vignettes/HDAnalyzeR.Rmd" class="external-link"><code>vignettes/HDAnalyzeR.Rmd</code></a></small>
      <div class="d-none name"><code>HDAnalyzeR.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="what-is-hdanalyzer">What is HDAnalyzeR?<a class="anchor" aria-label="anchor" href="#what-is-hdanalyzer"></a>
</h2>
<p>HDAnalyzeR is an R package developed to streamline and enhance
proteomics analysis, particularly for biomarker discovery from blood
plasma samples. It is developed by the Human Disease Blood Atlas group.
This vignette will guide you through the essential steps to use the
package from data loading to biomarker identification. Let’s get started
by loading the package!</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://kantonopoulos.github.io/HDAnalyzeR">HDAnalyzeR</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="how-to-get-help">How to get Help?<a class="anchor" aria-label="anchor" href="#how-to-get-help"></a>
</h2>
<p>If you need help with the package, you can use the <code>?</code>
operator to access the documentation for a specific function. For
example, to access the documentation for the
<code><a href="../reference/hd_initialize.html">hd_initialize()</a></code> function, you can run:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">?</span><span class="va">hd_initialize</span></span></code></pre></div>
<p>All package functions can be found in the <a href="https://hda1472.github.io/HDAnalyzeR/reference/index.html" class="external-link">HDAnalyzeR
Function Reference</a> page. For whole example workflows you can visit
any of the package articles.</p>
<p>This document introduces you to HDAnalyzeR’s basic set of tools, and
shows you how to analyze and identify biomarkers in a artificial dataset
of cancer blood plasma samples.</p>
</div>
<div class="section level2">
<h2 id="loading-the-data">Loading the Data<a class="anchor" aria-label="anchor" href="#loading-the-data"></a>
</h2>
<p>First, we load the package’s <code>example_data</code> and
<code>example_metadata</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">example_data</span><span class="op">)</span></span>
<span><span class="co">#&gt;      DAid    Sample  OlinkID UniProt  Assay           Panel        NPX</span></span>
<span><span class="co">#&gt; 1 DA00001 AML_syn_1 OID21311  Q9BTE6 AARSD1        Oncology  3.3903461</span></span>
<span><span class="co">#&gt; 2 DA00001 AML_syn_1 OID21280  P00519   ABL1        Oncology  2.7588517</span></span>
<span><span class="co">#&gt; 3 DA00001 AML_syn_1 OID21269  P09110  ACAA1        Oncology  1.7070090</span></span>
<span><span class="co">#&gt; 4 DA00001 AML_syn_1 OID20159  P16112   ACAN Cardiometabolic  0.0332709</span></span>
<span><span class="co">#&gt; 5 DA00001 AML_syn_1 OID20105  Q9BYF1   ACE2 Cardiometabolic  1.7553590</span></span>
<span><span class="co">#&gt; 6 DA00001 AML_syn_1 OID20124  Q15067  ACOX1 Cardiometabolic -0.9192835</span></span>
<span><span class="co">#&gt;   Assay_Warning QC_Warning PlateID</span></span>
<span><span class="co">#&gt; 1          PASS       PASS  Run001</span></span>
<span><span class="co">#&gt; 2          PASS       PASS  Run001</span></span>
<span><span class="co">#&gt; 3          PASS       PASS  Run001</span></span>
<span><span class="co">#&gt; 4          PASS       PASS  Run001</span></span>
<span><span class="co">#&gt; 5          PASS       PASS  Run001</span></span>
<span><span class="co">#&gt; 6          PASS       PASS  Run001</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">example_metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt;      DAid    Sample Disease   Stage Grade Sex Age  BMI Cohort</span></span>
<span><span class="co">#&gt; 1 DA00001 AML_syn_1     AML       2  &lt;NA&gt;   F  42 22.7   UCAN</span></span>
<span><span class="co">#&gt; 2 DA00002 AML_syn_2     AML Unknown  &lt;NA&gt;   M  69 33.1   UCAN</span></span>
<span><span class="co">#&gt; 3 DA00003 AML_syn_3     AML       2  &lt;NA&gt;   F  61 26.2   UCAN</span></span>
<span><span class="co">#&gt; 4 DA00004 AML_syn_4     AML Unknown  &lt;NA&gt;   M  54 28.1   UCAN</span></span>
<span><span class="co">#&gt; 5 DA00005 AML_syn_5     AML       2  &lt;NA&gt;   F  57 21.4   UCAN</span></span>
<span><span class="co">#&gt; 6 DA00006 AML_syn_6     AML Unknown  &lt;NA&gt;   M  86 33.9   UCAN</span></span></code></pre></div>
<p>We will initialize the HDAnalyzeR object with the data and metadata.
The <code><a href="../reference/hd_initialize.html">hd_initialize()</a></code> function will create an object that
will be used throughout the analysis. We will specify the sample ID,
variable name, and value name in order to correctly load and widen the
data. In this case, the data is in long format, so we set
<code>is_wide = FALSE</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hd_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_initialize.html">hd_initialize</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">example_data</span>, </span>
<span>                        metadata <span class="op">=</span> <span class="va">example_metadata</span>, </span>
<span>                        is_wide <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                        sample_id <span class="op">=</span> <span class="st">"DAid"</span>,</span>
<span>                        var_name <span class="op">=</span> <span class="st">"Assay"</span>,</span>
<span>                        value_name <span class="op">=</span> <span class="st">"NPX"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># See the data transformed into a wide format!</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hd_obj</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 101</span></span></span>
<span><span class="co">#&gt;   DAid    AARSD1  ABL1  ACAA1    ACAN   ACE2  ACOX1   ACP5   ACP6  ACTA2   ACTN4</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> DA00001   3.39  2.76  1.71   0.033<span style="text-decoration: underline;">3</span>  1.76  -<span style="color: #BB0000;">0.919</span> 1.54    2.15   2.81   0.742 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> DA00002   1.42  1.25 -<span style="color: #BB0000;">0.816</span> -<span style="color: #BB0000;">0.459</span>   0.826 -<span style="color: #BB0000;">0.902</span> 0.647   1.30   0.798 -<span style="color: #BB0000;">0.065</span><span style="color: #BB0000; text-decoration: underline;">9</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> DA00003  <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>      0.989  <span style="color: #BB0000;">NA</span>      0.330 1.37   <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     <span style="color: #BB0000;">NA</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> DA00004   3.41  3.38  1.69  <span style="color: #BB0000;">NA</span>       1.52  <span style="color: #BB0000;">NA</span>     0.841   0.582  1.70   0.108 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> DA00005   5.01  5.05  0.128  0.401  -<span style="color: #BB0000;">0.933</span> -<span style="color: #BB0000;">0.584</span> 0.026<span style="text-decoration: underline;">5</span>  1.16   2.73   0.350 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> DA00006   6.83  1.18 -<span style="color: #BB0000;">1.74</span>  -<span style="color: #BB0000;">0.156</span>   1.53  -<span style="color: #BB0000;">0.721</span> 0.620   0.527  0.772 <span style="color: #BB0000;">NA</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 90 more variables: ACY1 &lt;dbl&gt;, ADA &lt;dbl&gt;, ADA2 &lt;dbl&gt;, ADAM15 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ADAM23 &lt;dbl&gt;, ADAM8 &lt;dbl&gt;, ADAMTS13 &lt;dbl&gt;, ADAMTS15 &lt;dbl&gt;, ADAMTS16 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ADAMTS8 &lt;dbl&gt;, ADCYAP1R1 &lt;dbl&gt;, ADGRE2 &lt;dbl&gt;, ADGRE5 &lt;dbl&gt;, ADGRG1 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ADGRG2 &lt;dbl&gt;, ADH4 &lt;dbl&gt;, ADM &lt;dbl&gt;, AGER &lt;dbl&gt;, AGR2 &lt;dbl&gt;, AGR3 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   AGRN &lt;dbl&gt;, AGRP &lt;dbl&gt;, AGXT &lt;dbl&gt;, AHCY &lt;dbl&gt;, AHSP &lt;dbl&gt;, AIF1 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   AIFM1 &lt;dbl&gt;, AK1 &lt;dbl&gt;, AKR1B1 &lt;dbl&gt;, AKR1C4 &lt;dbl&gt;, AKT1S1 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   AKT3 &lt;dbl&gt;, ALCAM &lt;dbl&gt;, ALDH1A1 &lt;dbl&gt;, ALDH3A1 &lt;dbl&gt;, ALPP &lt;dbl&gt;, …</span></span></span></code></pre></div>
<blockquote>
<p>📓 In real-world scenarios, you would load your own data and metadata
files instead of using the example dataset.</p>
</blockquote>
<blockquote>
<p>⚠️ It is strongly recommended to use <code><a href="../reference/hd_initialize.html">hd_initialize()</a></code>
function to initialize an HDAnalyzeR object before starting any
analysis. This initialization step ensures that your input data and
metadata are correctly formatted and compatible with the package. While
it is possible to use HDAnalyzeR functions on data outside of an
HDAnalyzeR object, users must ensure that the data adheres to the
required structure: the first column must contain sample IDs, followed
by numeric columns in wide format. Additionally, the metadata must
include the same set of sample IDs. Proper initialization or careful
adherence to these requirements is crucial for accurate and efficient
analysis.</p>
</blockquote>
</div>
<div class="section level2">
<h2 id="quality-control-qc">Quality Control (QC)<a class="anchor" aria-label="anchor" href="#quality-control-qc"></a>
</h2>
<p><code><a href="../reference/hd_qc_summary.html">hd_qc_summary()</a></code> provides a comprehensive summary of the
input data and metadata.</p>
<p>Regarding the data, it will check the column types, calculate the
percentage of NAs in each column and row and plot their distributions,
calculate variable-variable (in this case protein-protein) correlations,
and create a heatmap of these correlations. Users can also specify the
threshold for reporting variable-variable correlations.</p>
<p>Regarding the metadata, it will check the column types and calculate
the percentage of NAs in each column and row exactly as before. It will
also create summary visualizations for the metadata variables like “Sex”
and “Age”.</p>
<p>Let’s run the QC summary and take a look at some results!</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_qc_summary.html">hd_qc_summary</a></span><span class="op">(</span><span class="va">hd_obj</span>, </span>
<span>                         variable <span class="op">=</span> <span class="st">"Disease"</span>, </span>
<span>                         palette <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Disease <span class="op">=</span> <span class="st">"cancers12"</span>, Sex <span class="op">=</span> <span class="st">"sex"</span><span class="op">)</span>, </span>
<span>                         cor_threshold <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>                         verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">qc_data</span><span class="op">$</span><span class="va">data_summary</span><span class="op">$</span><span class="va">na_col_hist</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qc_data</span><span class="op">$</span><span class="va">data_summary</span><span class="op">$</span><span class="va">cor_heatmap</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-7-2.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qc_data</span><span class="op">$</span><span class="va">metadata_summary</span><span class="op">$</span><span class="va">Sex</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-7-3.png" width="700"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">qc_data</span><span class="op">$</span><span class="va">metadata_summary</span><span class="op">$</span><span class="va">Age</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 6.06</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-7-4.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="imputation-and-dimensionality-reduction">Imputation and Dimensionality Reduction<a class="anchor" aria-label="anchor" href="#imputation-and-dimensionality-reduction"></a>
</h2>
<p>As we saw from the QC results, the data contains NAs and other issues
that need to be addressed. We will impute missing values using K-nearest
neighbors (KNN) with 5 neighbors via <code><a href="../reference/hd_impute_knn.html">hd_impute_knn()</a></code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hd_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_impute_knn.html">hd_impute_knn</a></span><span class="op">(</span><span class="va">hd_obj</span>, k <span class="op">=</span> <span class="fl">5</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hd_obj</span><span class="op">$</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 101</span></span></span>
<span><span class="co">#&gt;   DAid    AARSD1  ABL1  ACAA1    ACAN   ACE2  ACOX1   ACP5  ACP6 ACTA2   ACTN4</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> DA00001   3.39  2.76  1.71   0.033<span style="text-decoration: underline;">3</span>  1.76  -<span style="color: #BB0000;">0.919</span> 1.54   2.15  2.81   0.742 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> DA00002   1.42  1.25 -<span style="color: #BB0000;">0.816</span> -<span style="color: #BB0000;">0.459</span>   0.826 -<span style="color: #BB0000;">0.902</span> 0.647  1.30  0.798 -<span style="color: #BB0000;">0.065</span><span style="color: #BB0000; text-decoration: underline;">9</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> DA00003   3.70  2.85  2.27   0.989   0.184  0.330 1.37   0.876 1.33   0.570 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> DA00004   3.41  3.38  1.69   0.415   1.52   1.23  0.841  0.582 1.70   0.108 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> DA00005   5.01  5.05  0.128  0.401  -<span style="color: #BB0000;">0.933</span> -<span style="color: #BB0000;">0.584</span> 0.026<span style="text-decoration: underline;">5</span> 1.16  2.73   0.350 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> DA00006   6.83  1.18 -<span style="color: #BB0000;">1.74</span>  -<span style="color: #BB0000;">0.156</span>   1.53  -<span style="color: #BB0000;">0.721</span> 0.620  0.527 0.772  0.446 </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 90 more variables: ACY1 &lt;dbl&gt;, ADA &lt;dbl&gt;, ADA2 &lt;dbl&gt;, ADAM15 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ADAM23 &lt;dbl&gt;, ADAM8 &lt;dbl&gt;, ADAMTS13 &lt;dbl&gt;, ADAMTS15 &lt;dbl&gt;, ADAMTS16 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ADAMTS8 &lt;dbl&gt;, ADCYAP1R1 &lt;dbl&gt;, ADGRE2 &lt;dbl&gt;, ADGRE5 &lt;dbl&gt;, ADGRG1 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   ADGRG2 &lt;dbl&gt;, ADH4 &lt;dbl&gt;, ADM &lt;dbl&gt;, AGER &lt;dbl&gt;, AGR2 &lt;dbl&gt;, AGR3 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   AGRN &lt;dbl&gt;, AGRP &lt;dbl&gt;, AGXT &lt;dbl&gt;, AHCY &lt;dbl&gt;, AHSP &lt;dbl&gt;, AIF1 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   AIFM1 &lt;dbl&gt;, AK1 &lt;dbl&gt;, AKR1B1 &lt;dbl&gt;, AKR1C4 &lt;dbl&gt;, AKT1S1 &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   AKT3 &lt;dbl&gt;, ALCAM &lt;dbl&gt;, ALDH1A1 &lt;dbl&gt;, ALDH3A1 &lt;dbl&gt;, ALPP &lt;dbl&gt;, …</span></span></span></code></pre></div>
<p>After imputation, we will run Principal Component Analysis (PCA) via
<code><a href="../reference/hd_auto_pca.html">hd_auto_pca()</a></code> and Uniform Manifold Approximation and
Projection (UMAP) via <code><a href="../reference/hd_auto_umap.html">hd_auto_umap()</a></code> to check for outliers,
batch effects, and other potential issues. These are wrapper functions
that will run the necessary steps to perform PCA and UMAP, including
scaling the data and coloring the plots by metadata variables.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_auto_pca.html">hd_auto_pca</a></span><span class="op">(</span><span class="va">hd_obj</span>, </span>
<span>                       components <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                       plot_color <span class="op">=</span> <span class="st">"Disease"</span>, </span>
<span>                       plot_palette <span class="op">=</span> <span class="st">"cancers12"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pca_res</span><span class="op">$</span><span class="va">pca_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 11</span></span></span>
<span><span class="co">#&gt;   DAid     PC1     PC2    PC3    PC4    PC5    PC6    PC7    PC8     PC9    PC10</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> DA00… -<span style="color: #BB0000;">3.69</span>  -<span style="color: #BB0000;">4.32</span>   -<span style="color: #BB0000;">2.30</span>  -<span style="color: #BB0000;">3.15</span>  -<span style="color: #BB0000;">2.53</span>  -<span style="color: #BB0000;">2.78</span>  -<span style="color: #BB0000;">2.77</span>   0.603 -<span style="color: #BB0000;">3.20</span>   -<span style="color: #BB0000;">0.532</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> DA00…  4.17  -<span style="color: #BB0000;">2.70</span>    2.10  -<span style="color: #BB0000;">0.472</span> -<span style="color: #BB0000;">4.47</span>  -<span style="color: #BB0000;">1.96</span>  -<span style="color: #BB0000;">1.04</span>   1.51  -<span style="color: #BB0000;">0.229</span>  -<span style="color: #BB0000;">0.061</span><span style="color: #BB0000; text-decoration: underline;">4</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> DA00… -<span style="color: #BB0000;">3.31</span>   4.74    1.41   0.920 -<span style="color: #BB0000;">0.588</span>  0.316 -<span style="color: #BB0000;">0.103</span> -<span style="color: #BB0000;">0.441</span>  1.36    0.201 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> DA00… -<span style="color: #BB0000;">4.78</span>   0.447   1.44   0.134 -<span style="color: #BB0000;">1.15</span>  -<span style="color: #BB0000;">0.284</span>  0.279  3.34   0.045<span style="text-decoration: underline;">2</span> -<span style="color: #BB0000;">0.866</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> DA00… -<span style="color: #BB0000;">5.04</span>  -<span style="color: #BB0000;">3.69</span>    0.820 -<span style="color: #BB0000;">5.73</span>  -<span style="color: #BB0000;">0.697</span> -<span style="color: #BB0000;">3.87</span>  -<span style="color: #BB0000;">1.01</span>   1.96  -<span style="color: #BB0000;">0.970</span>  -<span style="color: #BB0000;">1.12</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> DA00…  0.429  0.071<span style="text-decoration: underline;">4</span> -<span style="color: #BB0000;">1.86</span>  -<span style="color: #BB0000;">7.80</span>   0.809 -<span style="color: #BB0000;">2.89</span>  -<span style="color: #BB0000;">0.733</span>  0.381 -<span style="color: #BB0000;">0.617</span>   1.32</span></span>
<span><span class="va">pca_res</span><span class="op">$</span><span class="va">pca_plot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_res</span><span class="op">$</span><span class="va">pca_loadings_plot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-9-2.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_res</span><span class="op">$</span><span class="va">pca_variance_plot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-9-3.png" width="700"></p>
<p>Let’s color UMAP by Sex instead of Disease to see if there are any
sex related differences.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">umap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_auto_umap.html">hd_auto_umap</a></span><span class="op">(</span><span class="va">hd_obj</span>, </span>
<span>                         plot_color <span class="op">=</span> <span class="st">"Sex"</span>, </span>
<span>                         plot_palette <span class="op">=</span> <span class="st">"sex"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">umap_res</span><span class="op">$</span><span class="va">umap_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 3</span></span></span>
<span><span class="co">#&gt;   DAid    UMAP1  UMAP2</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> DA00001  2.25 -<span style="color: #BB0000;">0.712</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> DA00002 -<span style="color: #BB0000;">2.23</span> -<span style="color: #BB0000;">1.45</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> DA00003  1.90  1.65 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> DA00004  2.30  1.24 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> DA00005  2.55 -<span style="color: #BB0000;">0.620</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> DA00006 -<span style="color: #BB0000;">1.32</span>  1.29</span></span>
<span><span class="va">umap_res</span><span class="op">$</span><span class="va">umap_plot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="biomarker-identification">Biomarker Identification<a class="anchor" aria-label="anchor" href="#biomarker-identification"></a>
</h2>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<p>We will run a differential expression analysis to identify potential
biomarkers. We will use <code><a href="../reference/hd_de_limma.html">hd_de_limma()</a></code> so that we will be
able to correct also for Sex and Age. This method will help us pinpoint
proteins that are significantly different between conditions. We will
run the analysis 3 times for different case classes against all other
classes each time. We will also run <code><a href="../reference/hd_plot_volcano.html">hd_plot_volcano()</a></code> to
visualize the results. We will present only the results for Acute
Myeloid Leukemia (AML).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_res_aml</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_de_limma.html">hd_de_limma</a></span><span class="op">(</span><span class="va">hd_obj</span>,</span>
<span>                          variable <span class="op">=</span> <span class="st">"Disease"</span>,</span>
<span>                          case <span class="op">=</span> <span class="st">"AML"</span>,</span>
<span>                          correct <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sex"</span>, <span class="st">"Age"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/hd_plot_volcano.html">hd_plot_volcano</a></span><span class="op">(</span>logfc_lim <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">de_res_aml</span><span class="op">$</span><span class="va">de_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 10</span></span></span>
<span><span class="co">#&gt;   Feature  logFC   CI.L  CI.R AveExpr     t  P.Value adj.P.Val     B Disease</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> AZU1     1.54   1.23   1.84   0.226  9.97 9.51<span style="color: #949494;">e</span><span style="color: #BB0000;">-22</span>  9.51<span style="color: #949494;">e</span><span style="color: #BB0000;">-20</span>  38.5 AML    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ADA      1.40   1.12   1.68   0.924  9.75 6.37<span style="color: #949494;">e</span><span style="color: #BB0000;">-21</span>  3.19<span style="color: #949494;">e</span><span style="color: #BB0000;">-19</span>  36.6 AML    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> APBB1IP  1.08   0.799  1.35  -<span style="color: #BB0000;">0.253</span>  7.62 1.04<span style="color: #949494;">e</span><span style="color: #BB0000;">-13</span>  3.45<span style="color: #949494;">e</span><span style="color: #BB0000;">-12</span>  20.3 AML    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> APEX1    1.45   1.07   1.82   0.515  7.56 1.50<span style="color: #949494;">e</span><span style="color: #BB0000;">-13</span>  3.74<span style="color: #949494;">e</span><span style="color: #BB0000;">-12</span>  19.9 AML    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> ANGPT1  -<span style="color: #BB0000;">1.63</span>  -<span style="color: #BB0000;">2.09</span>  -<span style="color: #BB0000;">1.16</span>   1.72  -<span style="color: #BB0000;">6.87</span> 1.65<span style="color: #949494;">e</span><span style="color: #BB0000;">-11</span>  3.29<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span>  15.3 AML    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> ARTN     0.826  0.572  1.08   0.423  6.38 3.60<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span>  6.00<span style="color: #949494;">e</span><span style="color: #BB0000;">- 9</span>  12.3 AML</span></span>
<span><span class="va">de_res_aml</span><span class="op">$</span><span class="va">volcano_plot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_res_crc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_de_limma.html">hd_de_limma</a></span><span class="op">(</span><span class="va">hd_obj</span>,</span>
<span>                          variable <span class="op">=</span> <span class="st">"Disease"</span>,</span>
<span>                          case <span class="op">=</span> <span class="st">"CRC"</span>,</span>
<span>                          correct <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sex"</span>, <span class="st">"Age"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_res_lungc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_de_limma.html">hd_de_limma</a></span><span class="op">(</span><span class="va">hd_obj</span>,</span>
<span>                            variable <span class="op">=</span> <span class="st">"Disease"</span>,</span>
<span>                            case <span class="op">=</span> <span class="st">"LUNGC"</span>,</span>
<span>                            correct <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Sex"</span>, <span class="st">"Age"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can also summarize the results via
<code><a href="../reference/hd_plot_de_summary.html">hd_plot_de_summary()</a></code>. In order to use this function we need
to store all the results in a list.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"AML"</span> <span class="op">=</span> <span class="va">de_res_aml</span>, </span>
<span>               <span class="st">"CRC"</span> <span class="op">=</span> <span class="va">de_res_crc</span>, </span>
<span>               <span class="st">"LUNGC"</span> <span class="op">=</span> <span class="va">de_res_lungc</span><span class="op">)</span></span>
<span></span>
<span><span class="va">de_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_plot_de_summary.html">hd_plot_de_summary</a></span><span class="op">(</span><span class="va">de_res</span>, class_palette <span class="op">=</span> <span class="st">"cancers12"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_summary</span><span class="op">$</span><span class="va">de_barplot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_summary</span><span class="op">$</span><span class="va">upset_plot_up</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-15-2.png" width="700"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_summary</span><span class="op">$</span><span class="va">upset_plot_down</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-15-3.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="machine-learning-classification-model">Machine Learning Classification Model<a class="anchor" aria-label="anchor" href="#machine-learning-classification-model"></a>
</h3>
<p>In addition to differential expression analysis, we will use a
regularized machine learning classification model (LASSO) to identify
significant features. This model will help us understand which protein
features are most predictive of the conditions being studied.</p>
<p>We will start by creating the data split with
<code><a href="../reference/hd_split_data.html">hd_split_data()</a></code> so we don’t have any data leakage between
train and test data and then we will run the model with
<code><a href="../reference/hd_model_rreg.html">hd_model_rreg()</a></code>. Once again, we will run the analysis 3
times but we will present only the results for AML.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hd_split</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_split_data.html">hd_split_data</a></span><span class="op">(</span><span class="va">hd_obj</span>, variable <span class="op">=</span> <span class="st">"Disease"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lasso_res_aml</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_model_rreg.html">hd_model_rreg</a></span><span class="op">(</span><span class="va">hd_split</span>, </span>
<span>                               variable <span class="op">=</span> <span class="st">"Disease"</span>,</span>
<span>                               case <span class="op">=</span> <span class="st">"AML"</span>,</span>
<span>                               mixture <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                               grid_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                               palette <span class="op">=</span> <span class="st">"cancers12"</span>,</span>
<span>                               verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lasso_res_aml</span><span class="op">$</span><span class="va">metrics</span></span>
<span><span class="co">#&gt; $accuracy</span></span>
<span><span class="co">#&gt; [1] 0.7959184</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $sensitivity</span></span>
<span><span class="co">#&gt; [1] 0.9166667</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $specificity</span></span>
<span><span class="co">#&gt; [1] 0.7851852</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $auc</span></span>
<span><span class="co">#&gt; [1] 0.9351852</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $confusion_matrix</span></span>
<span><span class="co">#&gt;           Truth</span></span>
<span><span class="co">#&gt; Prediction   0   1</span></span>
<span><span class="co">#&gt;          0 106   1</span></span>
<span><span class="co">#&gt;          1  29  11</span></span>
<span><span class="va">lasso_res_aml</span><span class="op">$</span><span class="va">roc_curve</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-16-1.png" width="700"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lasso_res_aml</span><span class="op">$</span><span class="va">probability_plot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-16-2.png" width="700"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lasso_res_aml</span><span class="op">$</span><span class="va">feat_imp_plot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-16-3.png" width="700"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lasso_res_crc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_model_rreg.html">hd_model_rreg</a></span><span class="op">(</span><span class="va">hd_split</span>, </span>
<span>                               variable <span class="op">=</span> <span class="st">"Disease"</span>,</span>
<span>                               case <span class="op">=</span> <span class="st">"CRC"</span>,</span>
<span>                               mixture <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                               grid_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                               palette <span class="op">=</span> <span class="st">"cancers12"</span>,</span>
<span>                               verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lasso_res_lungc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_model_rreg.html">hd_model_rreg</a></span><span class="op">(</span><span class="va">hd_split</span>, </span>
<span>                                 variable <span class="op">=</span> <span class="st">"Disease"</span>,</span>
<span>                                 case <span class="op">=</span> <span class="st">"LUNGC"</span>,</span>
<span>                                 mixture <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                 grid_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                 palette <span class="op">=</span> <span class="st">"cancers12"</span>,</span>
<span>                                 verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>We can get a visual summary of the results via
<code><a href="../reference/hd_plot_model_summary.html">hd_plot_model_summary()</a></code>. In order to use this function we
need to store the results in a list as before.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lasso_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"AML"</span> <span class="op">=</span> <span class="va">lasso_res_aml</span>, </span>
<span>                  <span class="st">"CRC"</span> <span class="op">=</span> <span class="va">lasso_res_crc</span>, </span>
<span>                  <span class="st">"LUNGC"</span> <span class="op">=</span> <span class="va">lasso_res_lungc</span><span class="op">)</span></span>
<span></span>
<span><span class="va">features_summary</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_plot_model_summary.html">hd_plot_model_summary</a></span><span class="op">(</span><span class="va">lasso_res</span>, class_palette <span class="op">=</span> <span class="st">"cancers12"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">features_summary</span><span class="op">$</span><span class="va">metrics_barplot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-20-1.png" width="700"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">features_summary</span><span class="op">$</span><span class="va">features_barplot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-20-2.png" width="700"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">features_summary</span><span class="op">$</span><span class="va">upset_plot_features</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-20-3.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="one-step-further">One step further<a class="anchor" aria-label="anchor" href="#one-step-further"></a>
</h2>
<p>The final step involves performing a pathway enrichment analysis to
get more biological information about the features. In this example, we
will run an Over-Representation Analysis for the up-regulated proteins
from differential expression, that are also identified as features by
the ML model. We will use the Gene Ontology (GO) database and show the
results only for AML. We will use <code><a href="../reference/hd_ora.html">hd_ora()</a></code> and
<code><a href="../reference/hd_plot_ora.html">hd_plot_ora()</a></code> to perform the analysis and visualize its
results.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract the proteins identified by both DE and Lasso</span></span>
<span><span class="va">de_proteins</span> <span class="op">&lt;-</span> <span class="va">de_res_aml</span><span class="op">$</span><span class="va">de_res</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">Feature</span><span class="op">)</span></span>
<span></span>
<span><span class="va">lasso_proteins</span> <span class="op">&lt;-</span> <span class="va">lasso_res_aml</span><span class="op">$</span><span class="va">features</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Scaled_Importance</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">Feature</span><span class="op">)</span></span>
<span></span>
<span><span class="va">intersect_proteins</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://generics.r-lib.org/reference/setops.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">de_proteins</span>, <span class="va">lasso_proteins</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Perform ORA with GO database and visualize results</span></span>
<span><span class="va">enrichment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_ora.html">hd_ora</a></span><span class="op">(</span><span class="va">intersect_proteins</span>, </span>
<span>                     database <span class="op">=</span> <span class="st">"GO"</span>, </span>
<span>                     ontology <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                     pval_lim <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span>  <span class="co"># We will use a higher p-value threshold for this example</span></span>
<span><span class="va">enrichment_plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hd_plot_ora.html">hd_plot_ora</a></span><span class="op">(</span><span class="va">enrichment</span><span class="op">)</span></span>
<span></span>
<span><span class="va">enrichment_plots</span><span class="op">$</span><span class="va">dotplot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichment_plots</span><span class="op">$</span><span class="va">treeplot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-21-2.png" width="700"></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichment_plots</span><span class="op">$</span><span class="va">cnetplot</span></span></code></pre></div>
<p><img src="HDAnalyzeR_files/figure-html/unnamed-chunk-21-3.png" width="700"></p>
<blockquote>
<p>📓 Remember that these data are a dummy-dataset with artificial data
and the results in this guide should not be interpreted as real results.
The purpose of this vignette is to show you how to use the package and
its functions.</p>
</blockquote>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] glmnet_4.1-9      Matrix_1.7-3      parsnip_1.3.2     recipes_1.3.1    </span></span>
<span><span class="co">#&gt; [5] dplyr_1.1.4       HDAnalyzeR_0.99.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RcppAnnoy_0.0.22        splines_4.5.1           ggplotify_0.1.2        </span></span>
<span><span class="co">#&gt;   [4] tibble_3.3.0            R.oo_1.27.1             hardhat_1.4.1          </span></span>
<span><span class="co">#&gt;   [7] rpart_4.1.24            sparsevctrs_0.3.4       lifecycle_1.0.4        </span></span>
<span><span class="co">#&gt;  [10] globals_0.18.0          lattice_0.22-7          MASS_7.3-65            </span></span>
<span><span class="co">#&gt;  [13] backports_1.5.0         SnowballC_0.7.1         magrittr_2.0.3         </span></span>
<span><span class="co">#&gt;  [16] limma_3.64.1            sass_0.4.10             rmarkdown_2.29         </span></span>
<span><span class="co">#&gt;  [19] jquerylib_0.1.4         yaml_2.3.10             ggtangle_0.0.7         </span></span>
<span><span class="co">#&gt;  [22] askpass_1.2.1           reticulate_1.42.0       cowplot_1.2.0          </span></span>
<span><span class="co">#&gt;  [25] DBI_1.2.3               RColorBrewer_1.1-3      lubridate_1.9.4        </span></span>
<span><span class="co">#&gt;  [28] purrr_1.1.0             R.utils_2.13.0          BiocGenerics_0.54.0    </span></span>
<span><span class="co">#&gt;  [31] yulab.utils_0.2.0       nnet_7.3-20             ipred_0.9-15           </span></span>
<span><span class="co">#&gt;  [34] lava_1.8.1              GenomeInfoDbData_1.2.14 IRanges_2.42.0         </span></span>
<span><span class="co">#&gt;  [37] embed_1.1.5             S4Vectors_0.46.0        enrichplot_1.28.2      </span></span>
<span><span class="co">#&gt;  [40] ggrepel_0.9.6           tokenizers_0.3.0        listenv_0.9.1          </span></span>
<span><span class="co">#&gt;  [43] tidytree_0.4.6          pheatmap_1.0.13         umap_0.2.10.0          </span></span>
<span><span class="co">#&gt;  [46] RSpectra_0.16-2         parallelly_1.45.0       pkgdown_2.1.3          </span></span>
<span><span class="co">#&gt;  [49] codetools_0.2-20        DOSE_4.2.0              tidyselect_1.2.1       </span></span>
<span><span class="co">#&gt;  [52] shape_1.4.6.1           aplot_0.2.8             UCSC.utils_1.4.0       </span></span>
<span><span class="co">#&gt;  [55] farver_2.1.2            stats4_4.5.1            jsonlite_2.0.0         </span></span>
<span><span class="co">#&gt;  [58] ggridges_0.5.6          survival_3.8-3          iterators_1.0.14       </span></span>
<span><span class="co">#&gt;  [61] systemfonts_1.2.3       foreach_1.5.2           ggnewscale_0.5.2       </span></span>
<span><span class="co">#&gt;  [64] tools_4.5.1             treeio_1.32.0           ragg_1.4.0             </span></span>
<span><span class="co">#&gt;  [67] Rcpp_1.1.0              glue_1.8.0              prodlim_2025.04.28     </span></span>
<span><span class="co">#&gt;  [70] gridExtra_2.3           xfun_0.52               qvalue_2.40.0          </span></span>
<span><span class="co">#&gt;  [73] tidyheatmaps_0.2.1      GenomeInfoDb_1.44.0     withr_3.0.2            </span></span>
<span><span class="co">#&gt;  [76] fastmap_1.2.0           openssl_2.3.3           digest_0.6.37          </span></span>
<span><span class="co">#&gt;  [79] timechange_0.3.0        R6_2.6.1                gridGraphics_0.5-1     </span></span>
<span><span class="co">#&gt;  [82] textshaping_1.0.1       GO.db_3.21.0            RSQLite_2.4.1          </span></span>
<span><span class="co">#&gt;  [85] R.methodsS3_1.8.2       UpSetR_1.4.0            utf8_1.2.6             </span></span>
<span><span class="co">#&gt;  [88] tidyr_1.3.1             generics_0.1.4          data.table_1.17.8      </span></span>
<span><span class="co">#&gt;  [91] class_7.3-23            httr_1.4.7              htmlwidgets_1.6.4      </span></span>
<span><span class="co">#&gt;  [94] uwot_0.2.3              pkgconfig_2.0.3         gtable_0.3.6           </span></span>
<span><span class="co">#&gt;  [97] timeDate_4041.110       dials_1.4.0             blob_1.2.4             </span></span>
<span><span class="co">#&gt; [100] GPfit_1.0-9             workflows_1.2.0         XVector_0.48.0         </span></span>
<span><span class="co">#&gt; [103] furrr_0.3.1             clusterProfiler_4.16.0  janeaustenr_1.0.0      </span></span>
<span><span class="co">#&gt; [106] htmltools_0.5.8.1       fgsea_1.34.0            scales_1.4.0           </span></span>
<span><span class="co">#&gt; [109] Biobase_2.68.0          png_0.1-8               gower_1.0.2            </span></span>
<span><span class="co">#&gt; [112] ggfun_0.1.9             knitr_1.50              reshape2_1.4.4         </span></span>
<span><span class="co">#&gt; [115] nlme_3.1-168            org.Hs.eg.db_3.21.0     cachem_1.1.0           </span></span>
<span><span class="co">#&gt; [118] stringr_1.5.1           rsample_1.3.0           parallel_4.5.1         </span></span>
<span><span class="co">#&gt; [121] AnnotationDbi_1.70.0    desc_1.4.3              pillar_1.11.0          </span></span>
<span><span class="co">#&gt; [124] grid_4.5.1              vctrs_0.6.5             tune_1.3.0             </span></span>
<span><span class="co">#&gt; [127] vip_0.4.1               lhs_1.2.0               yardstick_1.3.2        </span></span>
<span><span class="co">#&gt; [130] evaluate_1.0.4          cli_3.6.5               compiler_4.5.1         </span></span>
<span><span class="co">#&gt; [133] rlang_1.1.6             crayon_1.5.3            tidytext_0.4.2         </span></span>
<span><span class="co">#&gt; [136] future.apply_1.20.0     labeling_0.4.3          plyr_1.8.9             </span></span>
<span><span class="co">#&gt; [139] forcats_1.0.0           fs_1.6.6                stringi_1.8.7          </span></span>
<span><span class="co">#&gt; [142] BiocParallel_1.42.1     Biostrings_2.76.0       lazyeval_0.2.2         </span></span>
<span><span class="co">#&gt; [145] GOSemSim_2.34.0         patchwork_1.3.1         bit64_4.6.0-1          </span></span>
<span><span class="co">#&gt; [148] future_1.58.0           ggplot2_3.5.2           KEGGREST_1.48.1        </span></span>
<span><span class="co">#&gt; [151] statmod_1.5.0           igraph_2.1.4            broom_1.0.8            </span></span>
<span><span class="co">#&gt; [154] memoise_2.0.1           bslib_0.9.0             ggtree_3.16.2          </span></span>
<span><span class="co">#&gt; [157] fastmatch_1.1-6         bit_4.6.0               gson_0.1.0             </span></span>
<span><span class="co">#&gt; [160] ape_5.8-1               DiceDesign_1.10</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Konstantinos Antonopoulos, Maria Bueno Alvez, Mathias Uhlen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Provided without <strong>any warranty</strong>.</p>
</div>

    </footer>
</div>





  
</body>
</html>
